svm/src/account_loader.rs

```diff
@@ -95,6 +98,260 @@ pub struct FeesOnlyTransaction {
     pub fee_details: FeeDetails,
 }
 
+#[cfg_attr(feature = "dev-context-only-utils", derive(Clone))]
+pub(crate) struct AccountLoader<'a, CB: TransactionProcessingCallback> {
+    pub program_cache: ProgramCacheForTxBatch,
+    account_overrides: Option<&'a AccountOverrides>,
+    account_cache: AHashMap<Pubkey, AccountCacheItem>,
+    callbacks: &'a CB,
+}
+impl<'a, CB: TransactionProcessingCallback> AccountLoader<'a, CB> {
+    pub fn new_with_account_cache_capacity(
+        callbacks: &'a CB,
+        account_overrides: Option<&'a AccountOverrides>,
+        program_cache: ProgramCacheForTxBatch,
+        capacity: usize,
+    ) -> AccountLoader<'a, CB> {
+        Self {
+            program_cache,
+            account_overrides,
+            account_cache: AHashMap::with_capacity(capacity),
+            callbacks,
+        }
+    }
+
+    pub fn load_account(
+        &mut self,
+        account_key: &Pubkey,
+        usage_pattern: AccountUsagePattern,
+    ) -> Option<LoadedTransactionAccount> {
+        let is_writable = usage_pattern == AccountUsagePattern::Writable;
+        let is_invisible = usage_pattern == AccountUsagePattern::ReadOnlyInvisible;
+
+        if let Some(cache_item) = self.account_cache.get_mut(account_key) {
+            // Non-builtin accounts owned by NativeLoader are never loaded into the cache.
+            let is_owned_by_loader = PROGRAM_OWNERS.iter().contains(cache_item.account.owner());
+
+            if !cache_item.inspected_as_writable {
+                // Inspecting a read-only account is harmless. Inspecting a writable
+                // account must be done before rent collection, and must never be done
+                // again in-batch, because inspection is intended to preserve the
+                // state prior to any writes.
+                self.callbacks.inspect_account(
+                    account_key,
+                    AccountState::Alive(&cache_item.account),
+                    is_writable,
+                );
+                cache_item.inspected_as_writable = cache_item.inspected_as_writable || is_writable;
+            }
+
+            if cache_item.account.lamports() == 0 {
+                None
+            } else if let Some(program) = (is_invisible && is_owned_by_loader)
+                .then_some(())
+                .and_then(|_| self.program_cache.find(account_key))
+            {
+                // NOTE if this account could be found in the program cache, we must
+                // substitute its size here, to perserve an oddity with transaction size
+                // calculations. However, we CANNOT skip the accounts cache check,
+                // in case the account was non-executable and modified in-batch.
```

Comment:
nit:
```suggestion
                // NOTE if this account could be found in the program cache, we must
                // substitute its size here to preserve an oddity with transaction size
                // calculations. However, we CANNOT skip the accounts cache check,
                // in case the account was non-executable and modified in-batch.
```